### PR TITLE
Handle escape, unicode, and hex  in `tokenize_escaped_single_quoted_string`

### PR DESCRIPTION
**Pull request by JasonLi-cn**
Related to https://github.com/apache/arrow-datafusion/pull/9268

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Thank you @JasonLi-cn  -- sorry for the delay in review. I think this PR does more than refactor the function, so I changed the name of the PR to reflect what I think is different

I also changed the API a little more to avoid some unecessary `pub (crate)`

Any chance you can take a look before I merge it to make sure I got it right?

**Comment by JasonLi-cn** [comment]
> Thank you @JasonLi-cn -- sorry for the delay in review. I think this PR does more than refactor the function, so I changed the name of the PR to reflect what I think is different
> 
> I also changed the API a little more to avoid some unecessary `pub (crate)`
> 
> Any chance you can take a look before I merge it to make sure I got it right?

Thank you for reviewing this PR, and I think your modification is necessary.

File: src/tokenizer.rs [hunk 1]
```diff
@@ -2139,4 +2236,68 @@ mod tests {
...
+        check_unescape(r"\x4", Some("\u{0004}"));
+        check_unescape(r"\x4L", Some("\u{0004}L"));
+        check_unescape(r"\x", Some("x"));
+        check_unescape(r"\xP", Some("xP"));
+        check_unescape(r"\x0", None);
+        check_unescape(r"\xCAD", None);
+        check_unescape(r"\xA9", None);
+
+        // octal byte value
+        check_unescape(r"\1", Some("\u{0001}"));
```
[the remark this answers is not in the document]
Comment (JasonLi-cn) [reply src/tokenizer.rs:2288]: These existing tests have been compared with Pg and are consistent with Pg. As for the test for 'multi character strings', I can add some more.

**Comment by alamb** [comment]
Thanks again @JasonLi-cn

**Pull request merged**